frmts/vrt/vrtdataset.cpp

```diff
@@ -1044,6 +1044,121 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
+                            CPLStrdup(strstr(papszSubdatasets[j], "=") + 1);
+                        GDALSubdatasetInfoH info =
+                            GDALGetSubdatasetInfo(pszSubdatasetSource);
+                        char *component =
+                            info
+                                ? GDALSubdatasetInfoGetSubdatasetComponent(info)
+                                : nullptr;
+
+                        bFound = component && EQUAL(pszValue, component);
+                        bFound_subdataset = true;
+                        CPLFree(component);
+                        GDALDestroySubdatasetInfo(info);
+                        if (bFound)
+                        {
+                            poSrcDS->ReleaseRef();
+                            poSrcDS = GDALDataset::Open(
+                                pszSubdatasetSource, GDAL_OF_RASTER,
+                                aosAllowedDrivers.List(), aosOpenOptions.List(),
+                                nullptr);
+                            CPLFree(pszSubdatasetSource);
+                            break;
+                        }
+                        else
+                        {
+                            CPLFree(pszSubdatasetSource);
+                        }
+                    }
+
+                    if (!bFound)
+                    {
+                        CPLError(CE_Failure, CPLE_IllegalArg,
+                                 "'subdataset' option should be be a valid "
+                                 "subdataset component name");
+                        poSrcDS->ReleaseRef();
+
+                        CPLFree(pszKey);
+                        return nullptr;
+                    }
+                }
+            }
+
+            if (EQUAL(pszKey, "subdataset_n"))
+            {
+                if (bFound_subdataset)
+                {
+                    CPLError(CE_Failure, CPLE_IllegalArg,
+                             "'subdataset_n' is mutually exclusive with option "
+                             "'subdataset'");
+                    poSrcDS->ReleaseRef();
+
+                    CPLFree(pszKey);
+                    return nullptr;
+                }
+                if (nSubdatasets > 0)
+                {
+                    int iSubdataset = atoi(pszValue);
+                    if (iSubdataset < 1 || iSubdataset > (nSubdatasets) / 2)
+                    {
+                        CPLError(
+                            CE_Failure, CPLE_IllegalArg,
+                            "'subdataset_n' option should indicate a valid "
+                            "subdataset component index (starting with 1)");
+                        CPLFree(pszKey);
+                        poSrcDS->ReleaseRef();
+                        return nullptr;
+                    }
+                    char *pszSubdatasetSource = CPLStrdup(
+                        strstr(papszSubdatasets[(iSubdataset - 1) * 2], "=") +
```

Comment:
- check strstr() for null
- assign to a std::string to avoid CPLStrdup()/CPLFree()